### ISSUE TITLE
Unable to run model.generate() using converted model

### ISSUE THREAD

**Issue by Zkli-hub** [astramind-ai/Mixture-of-depths#7]
I find that I cannot run the generate() function to inference inputs using the converted model, can you help me?

Here is the error:
![image](https://github.com/astramind-ai/Mixture-of-depths/assets/55663065/79cfc19e-3d82-4463-9133-21c349b3e05b)

**Comment by Zkli-hub** [comment]
@mlinmg  Seems that it still doesn't work even if I use the latest version and explicitly set model.eval().
You can reproduce it easily in this notebook which is forked from your latest repo version: https://github.com/Zkli-hub/Mixture-of-depths-test/blob/main/MoD/modeling/models/test.ipynb
And initiliaze the model using your own pretrained hf weight path.

**Comment by mlinmg** [comment]
I don't have a pretrained mod model atm, and it keeps giving me a nan related error, I'll do a small train those days

**Comment by Zkli-hub** [comment]
> I don't have a pretrained mod model atm, and it keeps giving me a nan related error, I'll do a small train those days

Thanks for your reply! Actually, you don't need a pretrained mod model to reproduce this. Just use the converted mod model from a base llama2-7b is okay. I suppose that it is not the problem of pretrained weight but the code. In my provided notebook, you can initialize the model using:
`model = LlamaMoDForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf")`
`tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf")`
`model.eval()`
And then you can find it still unable to run model.generate()

**Comment by Zkli-hub** [comment]
> I don't have a pretrained mod model atm, and it keeps giving me a nan related error, I'll do a small train those days

I think I know where's the problem. For the eval mode, the model generates output autoregressively which is a different logic from the training mode. It generates token one by one and there will be only one input token once.
I logged the intermediate variables of original llama model. The first figure shows the log of prepared input while the second shows the log of autoregressive inference stage which is a one by one input. To solve it, I think you can refer to the original paper and they listed two approach for the routing logic during the inference stage.
<img width="572" alt="image" src="https://github.com/astramind-ai/Mixture-of-depths/assets/55663065/24452c57-9a01-470e-a1ae-599e527217e4">
<img width="383" alt="image" src="https://github.com/astramind-ai/Mixture-of-depths/assets/55663065/2a7c06b6-2bef-4716-801f-8abe30bd7b10">

**Comment by mlinmg** [comment]
Actually there were couple of problems, the first one is cache managemnt, since when the router skip an entire block the cache get lost.
Also with llama3, is very tricky to not get NaN outputs while finetuning a model. I've seen that fa2 works way more reliably than sdpa, where it goes to NaN most of the time.
All issues should be addressed with 1.2.0 now

**Comment by ZHQSimon** [comment]
> > I don't have a pretrained mod model atm, and it keeps giving me a nan related error, I'll do a small train those days
> 
> I think I know where's the problem. For the eval mode, the model generates output autoregressively which is a different logic from the training mode. It generates token one by one and there will be only one input token once. I logged the intermediate variables of original llama model. The first figure shows the log of prepared input while the second shows the log of autoregressive inference stage which is a one by one input. To solve it, I think you can refer to the original paper and they listed two approach for the routing logic during the inference stage. <img alt="image" width="572" src="https://private-user-images.githubusercontent.com/55663065/338638039-24452c57-9a01-470e-a1ae-599e527217e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwNDk1OTksIm5iZiI6MTcxOTA0OTI5OSwicGF0aCI6Ii81NTY2MzA2NS8zMzg2MzgwMzktMjQ0NTJjNTctOWEwMS00NzBlLWExYWUtNTk5ZTUyNzIxN2U0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIyVDA5NDEzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0ZTgzYzJmZjY4MzE4NTUwNDkxZTIzNDM1NzA1YjMxNTk4NjY0NzU4Y2Q2MDlmZWI1NmVjOWVjMDY2MjhkMWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HVi8t7K3DqxaTntfi1KchLGuvkpxQ49_ERxXEamEw38"> <img alt="image" width="383" src="https://private-user-images.githubusercontent.com/55663065/338639703-2a7c06b6-2bef-4716-801f-8abe30bd7b10.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwNDk1OTksIm5iZiI6MTcxOTA0OTI5OSwicGF0aCI6Ii81NTY2MzA2NS8zMzg2Mzk3MDMtMmE3YzA2YjYtMmJlZi00NzE2LTgwMWYtOGFiZTMwYmQ3YjEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIyVDA5NDEzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3NWQwN2ZkM2VhNDQzZTZlN2FkZDFmNThhN2U0NjA5YmJmNTdjMzYzNmU1M2VjZWNmNDgyZGU4MTg3NmJlOTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oGEemBRP5pdEtNrjdBphHVK96SCxSe13DYkZYk1OlW8">



> > I don't have a pretrained mod model atm, and it keeps giving me a nan related error, I'll do a small train those days
> 
> I think I know where's the problem. For the eval mode, the model generates output autoregressively which is a different logic from the training mode. It generates token one by one and there will be only one input token once. I logged the intermediate variables of original llama model. The first figure shows the log of prepared input while the second shows the log of autoregressive inference stage which is a one by one input. To solve it, I think you can refer to the original paper and they listed two approach for the routing logic during the inference stage. <img alt="image" width="572" src="https://private-user-images.githubusercontent.com/55663065/338638039-24452c57-9a01-470e-a1ae-599e527217e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwNTE3MzYsIm5iZiI6MTcxOTA1MTQzNiwicGF0aCI6Ii81NTY2MzA2NS8zMzg2MzgwMzktMjQ0NTJjNTctOWEwMS00NzBlLWExYWUtNTk5ZTUyNzIxN2U0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIyVDEwMTcxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYyOGQyMWQ2ODZkZWE5NmY2NzA5MzIwNmE0NDA3NTIwOGUyOTBiMGM1Y2E4MjU4MGZlOTJjOTVmNjRkMTY5ODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.e20reumlwcGW5XGN1WS4S-6uS0woxPKygJ1x2r5IMEU"> <img alt="image" width="383" src="https://private-user-images.githubusercontent.com/55663065/338639703-2a7c06b6-2bef-4716-801f-8abe30bd7b10.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwNTE3MzYsIm5iZiI6MTcxOTA1MTQzNiwicGF0aCI6Ii81NTY2MzA2NS8zMzg2Mzk3MDMtMmE3YzA2YjYtMmJlZi00NzE2LTgwMWYtOGFiZTMwYmQ3YjEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIyVDEwMTcxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhNjJhMjZmMjZjZmU2YmM5ODc2ZWIzYzEzMWEzM2ZiM2Q4ODQ0ODRmNGQwMDU5ZWI5MDNmZjNhNmIyZjcyMTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3NHqwQHJiJuK_UJE4lOuwlD8ExdAKuZtXNU_599eMYs">

Excuse me, how did you get these correct intermediate variables?
